Better build and installation support

yanntm opened the pull request:
See issue https://github.com/ponder-lab/Optimize-Java-8-Streams-Refactoring/issues/208 for discussion.

yanntm committed:
use proper bundle dependencies instead of raw package

yanntm committed:
change paths to parent pom

yanntm committed:
refactoring parent pom id

yanntm committed:
a pom for the update site

yanntm committed:
update to match main ITSTools parent pom configuration

yanntm committed:
set versions to be SNAPSHOT based in pom

yanntm committed:
patching Manifest files for maven version build

yanntm committed:
revised versioning

yanntm committed:
more version poke

yanntm committed:
feature for maven

yanntm committed:
preparing travis build

yanntm committed:
.yml typos

yanntm committed:
.yml typos

yanntm committed:
patch travis path

yanntm committed:
switch to plugin (was fragment) + other changes

yanntm committed:
use repos identified by @OlivierBarbier

yanntm committed:
resolve on our newly built artifacts

yanntm committed:
new repo for metrics dependency

yanntm committed:
this project should be a fragment

yanntm committed:
add missing dependency

yanntm committed:
root project just to ease checkout

yanntm committed:
problems seem related to java version

yanntm committed:
skip coverage measures

yanntm committed:
a new "dev" feature for eval+test

yanntm committed:
new child modules

yanntm committed:
remove Java 11

yanntm committed:
discarding artifacts that should not live in the repo

yanntm committed:
drop discovery of original cuny update site

yanntm committed:
wrong plugin : citytech->ponder-lab

yanntm committed:
grab objenesis from maven central rather than embed

yanntm committed:
grab org.apache.commons.csv from Nexus

yanntm committed:
drop lib/ folder declaration

yanntm committed:
add org.objenesis dependency

yanntm committed:
discarded lib folder

yanntm committed:
more removal of lib/ folders

yanntm committed:
add dep to org.apache.commons.csv

yanntm committed:
let apache commons csv be part of the feature

yanntm committed:
let the "dev" feature also install pde+jdt source

yanntm committed:
SubProgressMonitor deprecated => SubMonitor

khatchad commented on edu.cuny.hunter.streamrefactoring.core/META-INF/MANIFEST.MF:
```diff
@@ -22,24 +18,18 @@ Require-Bundle: org.eclipse.ltk.core.refactoring,
  edu.cuny.hunter.streamrefactoring.core.refactorings,
  edu.cuny.hunter.streamrefactoring.core.utils,
  edu.cuny.hunter.streamrefactoring.core.wala
-Import-Package: com.ibm.safe.controller,
```
What's wrong with `Import-Package`?

yanntm replied:
"Import-Package" means that this particular package should exist in the platform, but you are not specifying the plugin that provides it. 
So it is a weakly declared dependency, in the sense you cannot actually import the package unless it is present in your platform (i.e. it has been installed). 
Moving this dependency declaration to a "Require-Bundle" means that maven and eclipse p2 both understand the dependency and will deal with it appropriately (i.e. by deploying the plugin you depend on, and its own dependencies). So for a "managed" build it is recommended not to use "Import-Package", I removed mostly all of those.

The problem is then to find a "bundle" or plugin that provides the required package. We can resolve such artifacts in the update sites we provide to maven :  https://github.com/ponder-lab/Optimize-Java-8-Streams-Refactoring/pull/209/files/b7f6acf331a63a36a04345feff1837490ef36337#diff-b45a7fd496ea4028cf723e4f5ecfcd38R108-R153

In this place I put the WALA and SAFE repositories, as well as point "orbit" which is an eclipse project providing quite a few common utilities packaged as bundles so we can use them in this kind of integration scenario. e.g. this is where we resolve dependencies on apache stuff https://download.eclipse.org/tools/orbit/downloads/drops/R20200529191137/ 
For instance this is where we now resolve org.apache.commons.csv dependencies (instead of packaging them ourselves in .eval plugin)

khatchad replied:
Oh, interesting. I didn't know about that. Thanks!

khatchad commented on .project:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
What is the advantage of having the root directory as an Eclipse project?

khatchad commented on edu.cuny.hunter.streamrefactoring.core/META-INF/MANIFEST.MF:
```diff
@@ -22,24 +18,18 @@ Require-Bundle: org.eclipse.ltk.core.refactoring,
+ org.eclipse.text,
+ org.eclipse.ui,
+ org.eclipse.pde.core,
+ org.objenesis
```
How does this get here? I see that the jar is removed, but now it is being referenced as a plug-in. Where did the plug-in come from?

yanntm replied:
I guess you mean the objenesis dependency.

The plugin is resolved directly in the Nexus central : 
https://github.com/yanntm/Optimize-Java-8-Streams-Refactoring/blob/master/edu.cuny.hunter.streamrefactoring.parent/pom.xml#L149-L152

A lot of third-party artifacts are released there, among which is this objenesis dependency. https://mvnrepository.com/artifact/org.objenesis/objenesis 

This is not as good as grabbing it from Orbit, because Orbit is a "simultaneous release" of components that are tested and interact well together and with the version of eclipse platform we are targetting.  But it avoids at least repackaging our own, and lets us update automatically if a new version is released to maven central.  It certainly is better than having a copy of the jar in our lib/ folder.

khatchad commented:
I will keep this open for the time being.

yanntm commented:
Ok sure, I'm still around if you need help adopting parts of the changes.
Concerning #192 the dependency on WALA version is currently cited here : https://github.com/yanntm/Optimize-Java-8-Streams-Refactoring/blob/master/edu.cuny.hunter.streamrefactoring.parent/pom.xml#L119-L123
We point to "streams" branch.

Then in the feature definition : https://github.com/yanntm/Optimize-Java-8-Streams-Refactoring/blob/master/edu.cuny.hunter.streamrefactoring.feature/feature.xml#L236-L263
We also make sure to make the WALA plugin explicitly part of our feature (as well as e.g. objenesis and other dependencies clients should not be able to find if they only have their default update sites configured + our own update-site). This way the update site generated will contain this particular plugin, clients will resolve it right there.

This eases and speeds up installation w.r.t. to using the "other update sites to visit" feature that you can configure, because client just contacts us + eclipse release update sites. 
So we are mirroring in the update site our more esoteric dependencies, with versions that come from maven repositories : orbit or maven central.

This fulfills the packaging needs that were covered by e.g. nesting jars in the lib folder.

yanntm commented:
i.e. we dropped these declarations : https://github.com/ponder-lab/Optimize-Java-8-Streams-Refactoring/pull/209/files#diff-fb138605633d3351e8d255d666debfd9L225-L231

the resolution of dependencies is in our hands now not the client's, we do it with maven at build time and package everything we need.

yanntm committed:
copied in openjdk 8 jars

from Java 9 up, these files that are needed by Wala no longer exist in
the JRE.

This is an attempt to package them with the tool, see also
https://github.com/ponder-lab/Optimize-Java-8-Streams-Refactoring/issues/206

yanntm committed:
add an activator for storage of the jars

yanntm committed:
test for general Project

yanntm committed:
FIX if we are not in Java Perspective

Then the selection could contain generic IProject instead of
IJavaProject

yanntm committed:
collect embedded RT.jar and others

yanntm committed:
unpack "core" to get the jars seeable

yanntm committed:
Activator + API to get the embedded JRE

Still working on
https://github.com/ponder-lab/Optimize-Java-8-Streams-Refactoring/issues/206

yanntm committed:
solve issues with tests + activator

Encountered :
!MESSAGE The Plug-in Development Environment requires the PlatformAdmin
service to operate. Please install the compatibility fragment
'org.eclipse.osgi.compatibility.state'.
Jun 21, 2020 6:48:15 PM
edu.cuny.streamrefactoring.core.plugin.StreamRefactorCorePlugin

yanntm committed:
improve trace to allow diagnostic

yanntm committed:
compilation issue with newer wider API of ProgressMonitor

yanntm committed:
included features become feature dependencies

yanntm committed:
align to 2021-03

yanntm committed:
drop travis-ci, go to gh-actions instead

yanntm committed:
typo in folder name

yanntm committed:
skip tests we need headless

yanntm committed:
we need xvfb for these

khatchad commented on .travis.yml:
```diff
@@ -1,12 +1,43 @@
+    
+script:
+ - cd edu.cuny.hunter.streamrefactoring.parent && mvn -e generate-sources install && cd -  
+ - du -sh edu.cuny.hunter.streamrefactoring.updatesite/target/repository
```
I also saw this in the GH actions build ...

yanntm replied:
the du is just because sometimes upload might fail a bit silently if the files are too large, it also makes sure it's about the right size (some MB) so it was (maybe) built correctly

yanntm replied:
This travis file is obsolete, they stopped offering free service for open source projects so I now only use GH actions. So this file can be deleted it is overridden by the Gh workflow file.

yanntm commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,35 @@
+      with:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          BRANCH: gh-pages # The branch the action should deploy to.
+          FOLDER: edu.cuny.hunter.streamrefactoring.updatesite/target/repository # The folder the action should deploy.
```
yes, but if you're considering merging this I should update to "modern" versions for a few of these, the syntax has evolved a bit see 
https://github.com/lip6/ITS-commandline/commit/e782ddc1ddf90769990123dbed1705680eae8dc4#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721R45-R52
for an example
